enumerable.rb

```diff
@@ -0,0 +1,124 @@
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    my_array = is_a?(Range) ? to_a : self
+
+    my_array.size.times { |n| yield (my_array[n]) }
+    self
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    my_array = is_a?(Range) ? to_a : self
+
+    my_array.size.times { |n| yield my_array[n], n }
+    self
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    result_array = []
+    my_each { |n| result_array << n if yield(n) }
+    result_array
+  end
+
+  # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+  def my_all(arg = nil)
+    result = true
+
+    my_each do |i|
+      result = if arg
+                 arg.is_a?(Class) ? i.is_a?(arg) : i.to_s.match?(arg.to_s)
+               elsif block_given?
+                 yield i
+               else
+                 i == false || i.nil? ? false : true
+               end
+      return false unless result
+    end
+
+    result
+  end
+
+  def my_any(arg = nil)
+    result = false
+    my_each do |i|
+      result = if arg
+                 arg.is_a?(Class) ? i.is_a?(arg) : i.to_s.match?(arg.to_s)
+               elsif block_given?
+                 yield i
+               else
+                 i == false || i.nil? ? false : true
+               end
+      return true if result
+    end
+    result
+  end
+
+  def my_none(arg = nil)
+    result = true
+    my_each do |i|
+      result = if arg
+                 arg.is_a?(Class) ? i.is_a?(arg) : i.to_s.match?(arg.to_s)
+               elsif block_given?
+                 !yield i
+               else
+                 !i == false || i.nil? ? false : true
+               end
+      return true if result
+    end
+    result
+  end
+
+  def my_count(arg = nil)
+    counter = 0
+
+    if !arg.nil?
+      my_each { |i| counter += 1 if i == arg }
+    elsif block_given?
+      my_each { |i| counter += 1 if yield(i) }
+    else
+      counter = size
+    end
+    counter
+  end
+
+  def my_map(a_proc = nil)
+    return to_enum unless block_given?
+
+    map_array = []
+
+    if a_proc.nil?
+      my_each { |n| map_array << yield(n) }
+    else
+      my_each { |n| map_array << proc.call(n) }
+    end
+
+    map_array
+  end
```

Comment:
### #my_map 
* [ ] executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

Reply:
this is impossible, ruby won't let you pass both a block
![error](https://user-images.githubusercontent.com/51087315/83946658-eb637000-a81a-11ea-9edc-a0c5bebbaf6c.jpg)
 and a proc